cypress setup & loading the app

sadakchap opened the pull request:
## Type of PR

- [x] Feature

## Description
- cypress setup
- a simple test to load up the app  

## Related issue

 this [issue](https://github.com/reobin/vimcolorschemes/issues/254)

## QA Instructions

Since, you mentioned to load up app & test for anything present there, is it good to check for header logo content as
![image](https://user-images.githubusercontent.com/44117648/96311092-5a04d580-1026-11eb-8936-ceb7bd024dca.png)


## Added unit tests?

- [ ] no, because they aren't needed

## Added documentation?

- [ ] no

reobin commented on cypress/integration/home_page_spec.js:
```diff
@@ -0,0 +1,6 @@
+describe('The Home Page', () => {
+  it('successfully loads', () => {
+    cy.visit('/')
+    // cy.contains('vim')
```
Is this commented on purpose?

sadakchap replied:
yes, I needed to ask you, that line no. 3 only checks whether app loads up or not. As you mentioned earlier that test should also check "if it has something in it".

For that, I think simply checking the static content ( like  page header logo text content) would be good? We can check that in line no. 4.

Is this what we require? please correct me if I am wrong.

reobin commented on package.json:
```diff
@@ -21,6 +21,7 @@
     "gatsby-source-filesystem": "^2.3.34",
     "gatsby-source-mongodb": "^2.3.12",
     "gatsby-transformer-sharp": "^2.5.17",
+    "mocha": "^8.1.3",
```
Is this a necessary dependency to run cypress?

sadakchap replied:
No,I think that is not necessary. Thank you, I will remove that.

reobin commented:
Thanks @sadakchap ! 🎉

A couple of things:
- We should have an `npm script` set up in `package.json` to easily run the tests. See `"test": "jest --silent"`. I'm thinking `test::e2e` should be the name for the script.
- Some of the files generated by cypress should be added to the git ignore to avoid adding them to the repository. I think `cypress/videos`, `cypress/support` and `cypress/plugins` will be sufficient.

sadakchap commented:
> Thanks @sadakchap ! 🎉
> 
> A couple of things:
> 
> * We should have an `npm script` set up in `package.json` to easily run the tests. See `"test": "jest --silent"`. I'm thinking `test::e2e` should be the name for the script.
> * Some of the files generated by cypress should be added to the git ignore to avoid adding them to the repository. I think `cypress/videos`, `cypress/support` and `cypress/plugins` will be sufficient.

Yep @reobin ,

- I will add a `npm script` in package.json file
- ok, I will add `cypress/videos`, `cypress/support` and `cypress/plugins` to gitignore

sadakchap commented:
Hi @reobin , I have updated test , added a npm script as you have suggested and also added cypress auto generated files to gitignore. 
Please, take a look and let me know about changes.

reobin commented on cypress/fixtures/example.json:
```diff
@@ -0,0 +1,5 @@
+  "name": "Using fixtures to represent data",
+  "email": "hello@cypress.io",
+  "body": "Fixtures are a great way to mock data for responses to routes"
+}
```
Is this an auto-generated file as well? If so, it should be added to the gitignore.

Also, I noticed running a test that fails creates a `cypress/screenshots` directory, which should be added to the gitignore as well.

sadakchap replied:
Yes, it is auto-generated file and I will add that too to gitignore

reobin commented on package.json:
```diff
@@ -53,7 +54,8 @@
     "test": "jest --silent",
-    "test::verbose": "jest"
+    "test::verbose": "jest",
+    "test::e2e": "node_modules\\.bin\\cypress open"
```
For CI purpose, running `cypress run` will be sufficient.

We could have two scripts:
```json
{
  "test::e2e": "cypress run",
  "cypress::open": "cypress open"
}
```

The first one is more general and will be run on every PRs. The second will be used if necessary when developing locally.

sadakchap replied:
Yep!! actually that makes sense.

sadakchap replied:
But, just a thought 💭. @reobin 
If running test locally while developing, if the developer runs `cypress::open` locally and does not have cypress installed in the  system, but installed in particular that environment(/node_modules).

Then, it will give an error saying `cypress command is not recognized` .
Then, developer has to run like `node_modules\\.bin\\cypress open` running. Right?
Please, correct me if I am wrong because, this is first time I am using cypress

reobin replied:
Since the script is run by `npm`, it will use the `cypress` that is installed in the `node_modules`, so a user should not have any problem running `"cypress::open": "cypress open"` even if it is not installed on the machine directly.

sadakchap replied:
Ahh, right! I missed that. Thank you😃

reobin commented:
Small comments, then we're good to go 🌞

sadakchap commented:
@reobin , I tried updating it as you have suggested. Please take a look.

reobin commented:
Last thing @sadakchap 😬

Could you run `npm run format` and commit again? I think you skipped the husky pre-commit hook somehow

sadakchap commented:
> Last thing @sadakchap 😬
> 
> Could you run `npm run format` and commit again? I think you skipped the husky pre-commit hook somehow

I actually forgot to ran that command.🙄

reobin commented:
Looks good @sadakchap ! 🎉

I'll look into integrating this into the GitHub actions. Thanks so much!

@all-contributors please add @sadakchap for code and test

sadakchap commented:
> Looks good @sadakchap ! 🎉
> 
> I'll look into integrating this into the GitHub actions. Thanks so much!
> 
> @all-contributors please add @sadakchap for code and test

Thank you @reobin for being so patient and helping me😄😄
